### PR TITLE
feat: signal we have wrapped fetch

### PR DESCRIPTION
**Pull request by pauldambra**
A customer asked how they could be sure network payload capture was ready before making an API request they wanted to see the content of

This was for a debug situation but it's a reasonable request. Let's have a callback when _session replay_ is ready which has a wider usefulness

I was going to add an `onX` callback but could never get the shape of it somewhere I was happy with. That's my instinct telling me something so I'm going to unwind the changes here and attack this differently separately

When we patch a network call we now add `__posthog_wrapped__` to it so someone can tell whether we have wrapped it.

We already add `__rrweb_original__` but if two systems were using the rrweb plugin we'd not know which was us

Includes. bunch of test tidiness improvements, and adds a test on network capture

Screenshot 2024-03-14 at 20 37 30

### CODE REVIEW

File: cypress/support/e2e.ts [hunk 1]
```diff
@@ -9,24 +9,36 @@ Cypress.on('window:before:load', (win) => {
 })
 
 beforeEach(() => {
-    cy.intercept('POST', '**/decide/*').as('decide')
-    cy.intercept('POST', '**/e/*').as('capture')
-    cy.intercept('POST', '**/ses/*').as('session-recording')
-    cy.intercept('GET', '**/surveys/*').as('surveys')
+    cy.intercept('POST', '/decide/*').as('decide')
```
Comment (pauldambra) [inline cypress/support/e2e.ts:12]: as far as i can tell `'/decide/*'` matches more than `'**/decide/*'` 

from console errors it appears `**/decide` does not match a call that is simple `/decide` (which the browser interprets as `{current base url}/decide`

File: cypress/support/e2e.ts [hunk 2]
```diff
@@ -9,24 +9,36 @@ Cypress.on('window:before:load', (win) => {
...
+    cy.intercept('POST', '/e/*', { status: 1 }).as('capture')
```
Comment (pauldambra) [inline cypress/support/e2e.ts:13]: default response removes console spam from the rate limiter

File: cypress/support/e2e.ts [hunk 3]
```diff
@@ -9,24 +9,36 @@ Cypress.on('window:before:load', (win) => {
...
+    cy.intercept('GET', '/surveys/*').as('surveys')
 
     cy.readFile('dist/array.full.js').then((body) => {
-        cy.intercept('**/static/array.full.js', { body })
+        cy.intercept('/static/array.full.js', { body })
+    })
+
+    cy.readFile('dist/array.full.js.map').then((body) => {
+        cy.intercept('/static/array.full.js.map', { body })
     })
```
Comment (pauldambra) [inline cypress/support/e2e.ts:23]: returning map files removes console spam from the browser's attempt to load sourcemaps

File: src/types.ts [hunk 4]
```diff
@@ -184,6 +184,13 @@ export interface SessionRecordingOptions {
     // our settings here only support a subset of those proposed for rrweb's network capture plugin
     recordHeaders?: boolean
     recordBody?: boolean
+    /**
+     * This callback is fired when network capture has started,
+     * it can be used for example to send API calls you want to ensure
+     * are captured by PostHog that could otherwise fire on page load
+     * _before_ PostHog is ready to capture them
+     */
+    onNetworkCaptureReady?: () => void
```
Comment (marandaneto) [inline src/types.ts:193]: What if we do something like this: https://github.com/PostHog/posthog-js-lite/blob/e004b602463df7ef4990f2958665c622b53be252/posthog-core/src/index.ts#L153-L155
Which is a callback that can emit different types of events, then we can type an `eventType` string such as `string: onNetworkCaptureReady | captureReady | featureFlagsReady`, etc, so we can slowly add more on top of it.

Comment (pauldambra) [reply src/types.ts:193]: Ooh I didn't know we had that... I think the customer that requested the feature has a way forward without this PR - that's way nicer though... 🤔

Comment (pauldambra) [reply src/types.ts]: Hmmm, one for rested monday brain I think. 

posthog-js has the contract that a listener is called  when it is registered which I can't figure out with tired friday brain :)

**Review by marandaneto — APPROVED** [review]

**Comment by pauldambra** [comment]
logged https://github.com/PostHog/posthog-js/issues/1093 because I don't want to lose sight of the good improvement that would offer

**Pull request merged**